src/object_async/hello_async.cpp

```diff
@@ -0,0 +1,227 @@
+  // therefore threads are busy
+  std::string do_expensive_work(bool louder, std::string name) {  
+
+      std::map<std::size_t,std::string> container;  
+      std::size_t work_to_do=100000;
+
+      for (std::size_t i=0;i<work_to_do;++i) {
+          container.emplace(i,std::to_string(i));
+      }  
+
+      for (std::size_t i=0;i<work_to_do;++i) {
+          std::string const& item = container[i];
+          if (item != std::to_string(i)) {
+
+              // AsyncHelloWorker's Execute function will take care of this error 
+              // and return it to js-world via callback
+              throw std::runtime_error("Uh oh, this should never happen");
+          }
+      }  
+
+      std::string result = "...threads are busy async bees...hello " + name;
+ 
+      if (louder)
+      {
+        result += "!!!!";
+      }
+     
+      return result;  
+
+  }  
+
+  // This is the worker running asynchronously and calling a user-provided callback when done.
+  // Consider storing all C++ objects you need by value or by shared_ptr to keep them alive until done.
+  // Nan AsyncWorker docs: https://github.com/nodejs/nan/blob/master/doc/asyncworker.md
+  struct AsyncHelloWorker : Nan::AsyncWorker {
+
+      using Base = Nan::AsyncWorker;  
+
+      AsyncHelloWorker(bool louder, std::string name, Nan::Callback* callback)
+          : Base(callback), result_{""}, louder_{louder}, name_{name} { }  
+
+      // The Execute() function is getting called when the worker starts to run.
+      // - You only have access to member variables stored in this worker.
+      // - You do not have access to Javascript v8 objects here.
+      void Execute() override {
+          try {
+              result_ = do_expensive_work(louder_, name_);
+          } catch (const std::exception& e) {
+              SetErrorMessage(e.what());
+          }
+      }  
+
+      // The HandleOKCallback() is getting called when Execute() successfully completed.
+      // - In case Execute() invoked SetErrorMessage("") this function is not getting called.
+      // - You have access to Javascript v8 objects again
+      // - You have to translate from C++ member variables to Javascript v8 objects
+      // - Finally, you call the user's callback with your results
+      void HandleOKCallback() override {
+          Nan::HandleScope scope;  
+
+          const auto argc = 2u;
+          v8::Local<v8::Value> argv[argc] = {Nan::Null(), Nan::New<v8::String>(result_).ToLocalChecked()};  
+
+          callback->Call(argc, argv);
+      }  
+
+      std::string result_;
+      const bool louder_;
+      std::string name_;
+  };  
+
+
+  NAN_METHOD(HelloObjectAsync::hello_async) {
+    // "info" comes from the NAN_METHOD macro, which returns differently
+    // according to the version of node
+    // Mention anything about "Unwrap"?
+    HelloObjectAsync* h = Nan::ObjectWrap::Unwrap<HelloObjectAsync>(info.Holder());
+
+    std::string name = h->name_;
```

Comment:
Since you're pointing out string copies here - not sure if it makes sense to use move-semantics throughout this example (e.g. to move the string into the async worker). Maybe it just makes it more complicated than it has to be..